src/main/java/de/tum/in/www1/artemis/service/scheduled/quiz/QuizScheduleService.java

```diff
@@ -341,87 +464,101 @@ public void processCachedQuizSubmissions() {
+                // Get fresh QuizExercise from DB
+                QuizExercise quizExercise = quizExerciseService.findOne(quizExerciseId);
                 // check if quiz has been deleted
                 if (quizExercise == null) {
-                    submissionHashMap.remove(quizExerciseId);
+                    log.debug("Remove quiz " + quizExerciseId + " from resultHashMap");
+                    cachedQuizExercises.remove(quizExerciseId);
+                    cachedQuiz.destroy();
                     continue;
                 }
 
-                // if quiz has ended, all submissions will be processed => we can remove the inner HashMap for this quiz
-                // if quiz hasn't ended, some submissions (those that are not submitted) will stay in HashMap => keep inner HashMap
-                Map<String, QuizSubmission> submissions;
-                if (quizExercise.isEnded()) {
-                    submissions = submissionHashMap.remove(quizExerciseId);
-                }
-                else {
-                    submissions = submissionHashMap.get(quizExerciseId);
+                // (Boolean wrapper is safe to auto-unbox here)
+                boolean hasEnded = quizExercise.isEnded();
+                // Note that those might not be true later on due to concurrency and a distributed system,
+                // do not rely on that for actions upon the whole set, such as clear()
+                boolean hasNewSubmissions = !cachedQuiz.getSubmissions().isEmpty();
+                boolean hasNewParticipations = !cachedQuiz.getParticipations().isEmpty();
+                boolean hasNewResults = !cachedQuiz.getResults().isEmpty();
+
+                // Skip quizzes with no cached changes
+                if (!hasNewSubmissions && !hasNewParticipations && !hasNewResults) {
+                    // Remove quiz if it is not scheduled for start
+                    if (hasEnded && cachedQuiz.getQuizStart().isEmpty()) {
+                        removeCachedQuiz(cachedQuiz);
+                    }
+                    continue;
                 }
 
-                int numberOfSubmittedSubmissions = saveQuizSubmissionWithParticipationAndResultToDatabase(quizExercise, submissions);
+                // Update cached exercise object (use the expensive operation upfront)
+                quizExercise = quizExerciseService.findOneWithQuestionsAndStatistics(quizExerciseId);
+                cachedQuiz.setExercise(quizExercise);
+                updateQuizExercise(quizExercise);
 
-                if (numberOfSubmittedSubmissions > 0) {
-                    log.info("Saved {} submissions to database in {} in quiz {}", numberOfSubmittedSubmissions, printDuration(start), quizExercise.getTitle());
-                }
-            }
+                // Save cached Submissions (this will also generate results and participations and place them in the cache)
+                long start = System.nanoTime();
```

Comment:
why is this coming so late now?

Reply:
I decided to move it there because it is first used as duration in `Saved {} submissions to database in {} in quiz {}`. I could also move it before `Long quizExerciseId = cachedQuiz.getId();` at the beginning of the loop. It would then of course measure more than saving the submissions to DB.